### DISCUSSION TITLE
Does math.log2 not work on uint8?

### DISCUSSION THREAD

**Discussion by MartinStuwe** [modular/modular#1736] [General]
I tried to learn how to implement the calculation of entropies with math.log2, and it seems like math.log2 doesn't work with DType.uint8.

I got:

log2[DType.uint8, 8](v1)
                    ^
/__w/modular/modular/Kernels/mojo/stdlib/math/math.mojo:1340:1: note: no viable expansions found
/__w/modular/modular/Kernels/mojo/stdlib/math/math.mojo:1355:42: note: call expansion failed - no concrete specializations
/__w/modular/modular/Kernels/mojo/stdlib/math/math.mojo:1256:1: note: no viable expansions found
/__w/modular/modular/Kernels/mojo/stdlib/math/math.mojo:1312:41: note: constraint failed: neginf only support on floating point types

I changed my code to use float32 and it worked.

from math import log2 
```
def main():

    var v1 = SIMD[DType.uint8, 8](1, 1, 1, 1, 1, 1, 1, 1)
    let v2 = log2[DType.uint8, 8](v1)
    print(v1)
    print(v2)
```

```
from math import log2 
def main():

    var v1 = SIMD[DType.float32, 8](1, 1, 1, 1, 1, 1, 1, 1)
    let v2 = log2[DType.float32, 8](v1)
    print(v1)
    print(v2)

```

I mean the log2 operation outputs floats, thus float32 makes sense, but shouldn't it be possible to use as input an uint8 vector?

Another question, why is log2 only defined for SIMD data types? Is it for performance reasons and thus all non logical math functions are defined with SIMD types?

Solved with:
```
from math import log2, mul

def main():
    
    var v1 = SIMD[DType.float32, 8](0.2, 0.1, 0.1, 0.1, 0.2, 0.1, 0.1, 0.1)

    let v2 = log2[DType.float32, 8](v1)
    var v3 = mul[DType.float32, 8](v1, v2)

    var entropy: Float32 = 0.0
    for i in range(8):
        entropy = entropy + v3[i]   

    entropy = -1 * entropy

    print("Entropy: ", entropy)
```

**Comment by richardkiss** [comment]
If you're just counting bits, maybe something in here would be helpful: https://docs.modular.com/mojo/stdlib/math/bit.html

**Reply by MartinStuwe** [reply]
Why bit counting? I used ones just for simplicity reasons to understand how to use the math functions.
Let's say there were also values of other integers, it seems like it's currently not possible to pass a SIMD vector of uint8 to log2, and based on the documentation it should work.

Or do you mean to compute log2 by getting the position of the most significant 1 of the binary representation?

But thanks, I will now just stick with float32 and try to continue to learn Mojo.
